--- gtk/gtkprintoperation-win32.c	2009-08-15 11:44:53.000000000 +0200
+++ gtk/gtkprintoperation-win32.c	2009-08-15 11:46:46.000000000 +0200
@@ -72,7 +72,7 @@
 
 static const GUID myIID_IPrintDialogCallback  = {0x5852a2c3,0x6530,0x11d1,{0xb6,0xa3,0x0,0x0,0xf8,0x75,0x7b,0xf9}};
 
-#if !defined (_MSC_VER) && !defined (MINGW64) && !defined (__MINGW64__)
+#if !defined (_MSC_VER) && !defined (__MINGW64_VERSION_MAJOR)
 #undef INTERFACE
 #define INTERFACE IPrintDialogCallback
 DECLARE_INTERFACE_ (IPrintDialogCallback, IUnknown)
